### ISSUE TITLE
Problem with Google Chrome and same page links 

### ISSUE THREAD

**Issue by MichelSuch** [nvaccess/nvda#5933]
On some sites, for example wikipedia articles, pressing space or enter on a same page link does not make NVDA jump to the appropriate paragraph.
If you have installed a navigation sounds extension, you hear that the link is actually activated, but NVDA does not follow.
This behaviour does not occur on all websites.

**Comment by jcsteh** [comment]
Technical: I haven't confirmed this yet, but my guess is that Chrome isn't firing scrollingStart events when the target is a named anchor (`<a name="...">` tag). This may even be because no accessibles get created for these tags. This needs to be investigated.

**Comment by feerrenrut** [comment]
I can reproduce this in Chrome Version 55.0.2883.87 m (64-bit), with NVDA Version: next-13761,fc9ed0a7 (Essentially equivalent to 2016.4).

`event_scrollingStart` in `gecko_ia2.py` is called and however `_handleScrollTo()` returns `false`. I'm trying to become a bit more familiar with this code so that I understand the implications of this. I'm not sure what `nextHandler` is or where it comes from.

**Comment by jcsteh** [comment]
You've spotted the key point here, which is that _handleScrollTo is
returning False. The question is why that's happening. I'd suggest logging
some info about the object that this is being called on. My hunch (based on
what you've found) is that Chrome is firing it on text leaf nodes (role
static text with 0 children), which NVDA can't handle. (Chrome probably
shouldn't do this, but let's confirm if I'm right first.)

For reference, nextHandler refers to the next event handler in the chain.
In this case, an object got a scrollingStart event, but that object was
part of a TreeInterceptor (specifically, a gecko_ia2 vbuf). Therefore, NVDA
lets the vbuf handle the event. If the vbuf decides to pass that event
further down the chain, it calls nextHandler, which points to the event
method in the NVDAObject itself (if any).

**Comment by feerrenrut** [comment]
## Summary
The role of `obj` is heading (value 40) and it has no children.

## Details

In the context of `_handleScrollTo(obj)`, the type of `obj` is `NVDAObjects.IAccessible.ia2Web.Ia2Web`

`_handleScrollTo` exits returning `false` because:

	_handleScrollTo got an exception
		Traceback (most recent call last):
		  File "browseMode.py", line 1363, in _handleScrollTo
		    scrollInfo = self.makeTextInfo(obj)
		  File "treeInterceptorHandler.py", line 143, in makeTextInfo
		    return self.TextInfo(self,position)
		  File "virtualBuffers\__init__.py", line 196, in __init__
		    super(VirtualBufferTextInfo,self).__init__(obj,position)
		  File "textInfos\offsets.py", line 285, in __init__
		    start,end=self._getOffsetsFromNVDAObject(position)
		  File "virtualBuffers\__init__.py", line 192, in _getOffsetsFromNVDAObject
		    raise LookupError
		LookupError

Which happens because the role is `ROLE_HEADING=40` which is not in the list of roles for which we should instead get the offset for the parent:

 - controlTypes.ROLE_LIST
 - controlTypes.ROLE_COMBOBOX
 - controlTypes.ROLE_GROUPING
 - controlTypes.ROLE_TREEVIEW
 - controlTypes.ROLE_TREEVIEWITEM

There are no children on `obj`.

This wouldn't happen if `_getOffsetsFromNVDAObjectInBuffer` returned, instead it throws a `LookupError` exception which is then caught and ignored:

	_getOffsetsFromNVDAObject got an exception
		Traceback (most recent call last):
		  File "virtualBuffers\__init__.py", line 184, in _getOffsetsFromNVDAObject
		    return self._getOffsetsFromNVDAObjectInBuffer(obj)
		  File "virtualBuffers\__init__.py", line 179, in _getOffsetsFromNVDAObjectInBuffer
		    return self._getOffsetsFromFieldIdentifier(docHandle,ID)
		  File "virtualBuffers\__init__.py", line 162, in _getOffsetsFromFieldIdentifier
		    raise LookupError
		LookupError

This exception happens because `NVDAHelper.localLib.VBuf_getControlFieldNodeWithIdentifier` does not return a valid node.

**Comment by jcsteh** [comment]
That is very strange indeed. We shouldn't need to walk to the parent
because the heading itself should be present in the buffer. Can you verify
that the target heading is present in browse mode (i.e. by jumping to it or
reading line by line)? Perhaps there are two objects for the same heading?
You could confirm that by checking the IA2UniqueID of the heading in the
document and the object being passed to _handleScrollTo.

**Comment by feerrenrut** [comment]
I had to edit my previous comment, it was incorrect and ambiguous. The role of obj is ROLE_STATIC_TEXT its parent has ROLE_HEADING.

Using chrome dev tools, the target (for the same page link) heading is as follows. Notice that the `id` is on the span within the heading:
```
<h2>
	<span class="mw-headline" id="Geography">Geography</span>
	<span class="mw-editsection">
		<span class="mw-editsection-bracket">[</span>
		<a href="/w/index.php?title=Monroe_Township,_Licking_County,_Ohio&amp;action=edit&amp;section=1" title="Edit section: Geography">edit</a>
		<span class="mw-editsection-bracket">]</span>
	</span>
</h2>
```

In "Accessibility Viewer" there are the following nodes:

 - Heading level 2: "Geography [ edit ]"
   - Text: "Geography"
   - Text: "[ "
   - Link: "edit"
     - Text: "edit"
   - Text: " ]"

With NVDA running I am able to arrow down to the heading, which is read out. Then when I press `nvda+F1` I get the following:
```
Developer info for navigator object:
name: u'Geography[edit]'
role: ROLE_HEADING
states: 
isFocusable: False
hasFocus: False
Python object: <NVDAObjects.IAccessible.ia2Web.Ia2Web object at 0x05876050>
Python class mro: (<class 'NVDAObjects.IAccessible.ia2Web.Ia2Web'>, <class 'NVDAObjects.IAccessible.IAccessible'>, <class 'NVDAObjects.window.Window'>, <class 'NVDAObjects.NVDAObject'>, <class 'baseObject.ScriptableObject'>, <class 'baseObject.AutoPropertyObject'>, <type 'object'>)
description: None
location: (2183, 498, 1282, 29)
value: u''
appModule: <'appModuleHandler' (appName u'chrome', process ID 6568) at address 5729450>
appModule.productName: u'Google Chrome'
appModule.productVersion: u'55.0.2883.87'
TextInfo: <class 'NVDAObjects.IAccessible.IA2TextTextInfo'>
windowHandle: 591312L
windowClassName: u'Chrome_RenderWidgetHostHWND'
windowControlID: 922265728
windowStyle: 1445986304
windowThreadID: 10548
windowText: u'Chrome Legacy Window'
displayText: u''
IAccessibleObject: <POINTER(IAccessible2) ptr=0x9bd5ac at 57e1ad0>
IAccessibleChildID: 0
IAccessible event parameters: windowHandle=591312, objectID=-4, childID=-2545
IAccessible accName: u'Geography[edit]'
IAccessible accRole: u'h2'
IAccessible accState: STATE_SYSTEM_HOTTRACKED, STATE_SYSTEM_VALID (128)
IAccessible accDescription: None
IAccessible accValue: u''
IAccessible2 windowHandle: 591312
IAccessible2 uniqueID: -2545
IAccessible2 role: IA2_ROLE_HEADING
IAccessible2 states: IA2_STATE_OPAQUE (1024)
IAccessible2 attributes: u'display:block;tag:h2;xml-roles:heading;level:2;'
```

When I print the `devInfo` for the obj passed into `_handleScrollTo`:

```
name: u'1 Geography'
name: u'Geography'
role: ROLE_STATICTEXT
states: 
isFocusable: False
hasFocus: False
Python object: <NVDAObjects.IAccessible.ia2Web.Ia2Web object at 0x059FA970>
Python class mro: (<class 'NVDAObjects.IAccessible.ia2Web.Ia2Web'>, <class 'NVDAObjects.IAccessible.IAccessible'>, <class 'NVDAObjects.window.Window'>, <class 'NVDAObjects.NVDAObject'>, <class 'baseObject.ScriptableObject'>, <class 'baseObject.AutoPropertyObject'>, <type 'object'>)
description: None
location: (2183, 90, 102, 25)
value: u''
appModule: <'appModuleHandler' (appName u'chrome', process ID 6568) at address 5729450>
appModule.productName: u'Google Chrome'
appModule.productVersion: u'55.0.2883.87'
TextInfo: <class 'NVDAObjects.IAccessible.IA2TextTextInfo'>
windowHandle: 591312L
windowClassName: u'Chrome_RenderWidgetHostHWND'
windowControlID: 922265728
windowStyle: 1445986304
windowThreadID: 10548
windowText: u'Chrome Legacy Window'
displayText: u''
IAccessibleObject: <POINTER(IAccessible2) ptr=0x92c30c at 59fe8f0>
IAccessibleChildID: 0
IAccessible event parameters: windowHandle=591312, objectID=-4, childID=-2865
IAccessible accName: u'Geography'
IAccessible accRole: ROLE_SYSTEM_STATICTEXT
IAccessible accState:  (0)
IAccessible accDescription: None
IAccessible accValue: u''
IAccessible2 windowHandle: 591312
IAccessible2 uniqueID: -2865
IAccessible2 role: ROLE_SYSTEM_STATICTEXT
IAccessible2 states: IA2_STATE_OPAQUE (1024)
IAccessible2 attributes: u'display:inline;'
```

It seems like the span with the `id` is being collapsed and is not present. This seems to be a chrome bug, events should not be raised with objects that have been collapsed.

In the mean time we can work around this issue by looking at the parent for objects with roles equal to ROLE_STATIC_TEXT. We will need to consider the impact of this a little first though.

EG:
```
def _getOffsetsFromNVDAObject(self, obj):
	while True:
		try:
			return self._getOffsetsFromNVDAObjectInBuffer(obj)
		except LookupError:
			pass
		isStaticTextRole = (obj.role == controlTypes.ROLE_STATICTEXT)
		# Interactive list/combo box/tree view descendants aren't rendered into the buffer, 
		# even though they are still considered part of it.
		# Use the container in this case.
		obj = obj.parent
		acceptedParentRoles = (controlTypes.ROLE_LIST, controlTypes.ROLE_COMBOBOX, controlTypes.ROLE_GROUPING,
			controlTypes.ROLE_TREEVIEW, controlTypes.ROLE_TREEVIEWITEM)
		if not obj or (not isStaticTextRole and obj.role not in acceptedParentRoles):
			break
	raise LookupError
```

**Comment by jcsteh** [comment]
Okay, so my hunch was correct: they're firing events on text leaf nodes. I
don't think this is correct. I think we should file a bug against Chrome
first and see if they can fix it before working around it.

**Comment by jcsteh** [comment]
There was already an issue open for this, so I commented on it with technical details and a test case: [https://bugs.chromium.org/p/chromium/issues/detail?id=37721](Chromium issue 37721).

**Comment by jcsteh** [comment]
I've confirmed that this is now fixed in Chrome Canary: Version 59.0.3068.1 (Official Build) canary (64-bit).

**Comment by zstanecic** [comment]
well, how  google releases often his stable versions of the browser? as 
i know, canary is not the stable one, may contain crashes and it’s unstable


W dniu 12.04.2017 o 08:43, James Teh pisze:
